File: app/src/main/java/torille/fi/lurkforreddit/search/SearchPresenter.java
package torille.fi.lurkforreddit.search;

import android.support.annotation.NonNull;
import android.support.v4.util.Pair;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;
import torille.fi.lurkforreddit.data.RedditRepository;
import torille.fi.lurkforreddit.data.models.view.SearchResult;

/**
 * Created by eva on 3/20/17.
 */

public class SearchPresenter implements SearchContract.Presenter<SearchContract.View> {

    private final RedditRepository mRedditRepository;
    private SearchContract.View mSearchView;
    private String searchAfter;
    private String searchQuery;
    private final CompositeDisposable disposables = new CompositeDisposable();

    @Inject
    SearchPresenter(RedditRepository mRedditRepository) {
        this.mRedditRepository = mRedditRepository;
    }

    @Override
    public void searchSubreddits(@NonNull String query) {
        this.searchQuery = query;
        mSearchView.clearResults();
        mSearchView.showProgressbar();
        disposables.add(mRedditRepository.getSearchResults(query)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(new DisposableObserver<Pair<String, List<SearchResult>>>() {
                    @Override
                    public void onNext(@io.reactivex.annotations.NonNull Pair<String, List<SearchResult>> resultPair) {
                        searchAfter = resultPair.first;
                        mSearchView.showResults(resultPair.second);
                    }

                    @Override
                    public void onError(@io.reactivex.annotations.NonNull Throwable e) {
                        Timber.e(e);
                        mSearchView.showError(e.toString());
                    }

                    @Override
                    public void onComplete() {

                    }
                }));
    }

    @Override
    public void searchMoreSubreddits() {
        mSearchView.showProgressbar();
        disposables.add(mRedditRepository.getMoreSearchResults(searchQuery, searchAfter)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(new DisposableObserver<Pair<String, List<SearchResult>>>() {
                    @Override
                    public void onNext(@io.reactivex.annotations.NonNull Pair<String, List<SearchResult>> resultPair) {
                        searchAfter = resultPair.first;
                        mSearchView.showResults(resultPair.second);
                    }

                    @Override
                    public void onError(@io.reactivex.annotations.NonNull Throwable e) {
                        Timber.e(e);
                        mSearchView.showError(e.toString());
                    }

                    @Override
                    public void onComplete() {

                    }
                }));
    }

    @Override
    public void setView(@NonNull SearchContract.View view) {
        mSearchView = view;
    }

    @Override
    public void start() {

    }

    @Override
    public void dispose() {
        disposables.dispose();
    }
}

File: app/src/main/java/torille/fi/lurkforreddit/search/SearchPresenter.kt
package torille.fi.lurkforreddit.search

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver
import io.reactivex.observers.SafeObserver
import io.reactivex.schedulers.Schedulers
import io.reactivex.subscribers.DisposableSubscriber
import io.reactivex.subscribers.ResourceSubscriber
import timber.log.Timber
import torille.fi.lurkforreddit.data.RedditRepository
import torille.fi.lurkforreddit.data.models.view.SearchResult
import javax.inject.Inject

class SearchPresenter @Inject
internal constructor(private val redditRepository: RedditRepository) : SearchContract.Presenter {

    private var searchView: SearchContract.View? = null
    private var searchAfter: String? = null
    private var searchQuery: String? = null
    private val disposables = CompositeDisposable()

    override fun searchSubreddits(query: String) {
        this.searchQuery = query
        searchView?.clearResults()
        searchView?.showProgressbar()
        disposables.add(redditRepository.getSearchResults(query)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(object : DisposableSubscriber<kotlin.Pair<String, List<SearchResult>>>() {
                    override fun onNext(resultPair: kotlin.Pair<String, List<SearchResult>>) {
                        searchAfter = resultPair.first
                        searchView?.showResults(resultPair.second)
                    }

                    override fun onError(@io.reactivex.annotations.NonNull e: Throwable) {
                        Timber.e(e)
                        searchView?.showError(e.toString())
                    }

                    override fun onComplete() {

                    }
                })
        )
    }

    override fun searchMoreSubreddits() {
        searchView?.showProgressbar()
        disposables.add(redditRepository.getMoreSearchResults(searchQuery!!, searchAfter!!)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(object : DisposableSubscriber<kotlin.Pair<String, List<SearchResult>>>() {
                    override fun onNext(resultPair: kotlin.Pair<String, List<SearchResult>>) {
                        searchAfter = resultPair.first
                        searchView?.showResults(resultPair.second)
                    }

                    override fun onError(@io.reactivex.annotations.NonNull e: Throwable) {
                        Timber.e(e)
                        searchView?.showError(e.toString())
                    }

                    override fun onComplete() {

                    }
                })
        )
    }

    override fun takeView(view: SearchContract.View) {
        searchView = view
    }

    override fun dropView() {
        disposables.dispose()
        searchView = null
    }


}
